pandas/core/array_algos/putmask.py

```diff
@@ -225,3 +237,84 @@ def setitem_datetimelike_compat(values: np.ndarray, num_set: int, other):
                 other = list(other)
 
     return other
+
+
+def putmask_flexible_ndarray(array: np.ndarray, mask, new):
+    """
+    Putmask implementation for ArrayManager putmask for ndarray.
+
+    Flexible version that will upcast if needed.
+    """
+    mask, noop = validate_putmask(array, mask)
+    assert not isinstance(new, (ABCIndex, ABCSeries, ABCDataFrame))
+
+    # if we are passed a scalar None, convert it here
+    if not array.dtype == "object" and is_valid_na_for_dtype(new, array.dtype):
+        new = na_value_for_dtype(array.dtype, compat=False)
+
+    if can_hold_element(array, new):
+        # error: Argument 1 to "putmask_without_repeat" has incompatible type
+        # "Union[ndarray, ExtensionArray]"; expected "ndarray"
+        putmask_without_repeat(array, mask, new)  # type: ignore[arg-type]
+        return array
+
+    elif noop:
+        return array
+
+    dtype, _ = infer_dtype_from(new)
+    if dtype.kind in ["m", "M"]:
+        array = array.astype(object)
+        # convert to list to avoid numpy coercing datetimelikes to integers
+        new = setitem_datetimelike_compat(
+            array, mask.sum(), new  # type: ignore[arg-type]
+        )
+        # putmask_smart below converts it back to array
+        np.putmask(array, mask, new)
+        return array
+
+    new_values = putmask_smart(array, mask, new)
+    return new_values
+
+
+def _coerce_to_target_dtype(array, new):
+    dtype, _ = infer_dtype_from(new, pandas_dtype=True)
+    new_dtype = find_common_type([array.dtype, dtype])
+    return array.astype(new_dtype, copy=False)
+
+
+def putmask_flexible_ea(array: ExtensionArray, mask, new):
```

Comment:
might help that there now exists EA._putmask?

Reply:
Let's review this part in https://github.com/pandas-dev/pandas/pull/44396 (I included that commit here so I could already fix some CoW issues, but so the actual refactor of putmask can be in a separate but). But yes.